src/Processors/Formats/Impl/Parquet/ParquetColumnReaderFactory.cpp

```diff
@@ -0,0 +1,1055 @@
+#include "ParquetColumnReaderFactory.h"
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDate32.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <DataTypes/DataTypeMap.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <DataTypes/DataTypesDecimal.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionHelpers.h>
+#include <Processors/Formats/Impl/Parquet/ParquetReader.h>
+#include <Processors/Formats/Impl/Parquet/RowGroupChunkReader.h>
+#include <Processors/Formats/Impl/Parquet/SelectiveColumnReader.h>
+#include <parquet/column_reader.h>
+#include <Common/ProfileEvents.h>
+
+namespace ProfileEvents
+{
+extern const Event ParquetFetchWaitTimeMicroseconds;
+}
+
+namespace ErrorCodes
+{
+extern const int PARQUET_EXCEPTION;
+extern const int NOT_IMPLEMENTED;
+extern const int LOGICAL_ERROR;
+}
+
+namespace DB
+{
+
+namespace
+{
+/// Create a column reader for a specific physical type and target type.
+/// \tparam physical_type physical type in parquet file
+/// \tparam target_type target type in ClickHouse
+/// \tparam dict is dictionary column or not
```

Comment:
I didn't read this in detail yet, but it seems strange to have separate readers for dict and non-dict cases. Column chunk may have some pages encoded with dictionary, then some pages encoded without dictionary (parquet writer falls back to non-dict encoding if the dictionary gets too big). So the `dict = true` reader must be able to switch to non-dict encoding on the fly. So why not use the same reader for both cases? If `dict = false`, initialize the reader in a state as if it already fell back from dict to non-dict encoding.